lib/common/schemas.c

```diff
@@ -1335,3 +1335,111 @@ pcmk__schemas_later_than(const char *name)
 
     return lst;
 }
+
+static void
+external_refs_in_schema(GList **list, const char *contents)
+{
+    xmlNode *ptr = string2xml(contents);
+    /* local-name()= is needed to ignore the xmlns= setting at the top of
+     * the XML file.  Otherwise, the xpath query will always return nothing.
+     */
+    xmlXPathObjectPtr xpathObj = xpath_search(ptr, "//*[local-name()='externalRef'] | //*[local-name()='include']");
+
+    for (int i = 0; i < numXpathResults(xpathObj); i++) {
+        xmlNode *result = getXpathResult(xpathObj, i);
+        const char *href = NULL;
+
+        if (result == NULL) {
+            continue;
+        }
+
+        href = crm_element_value(result, "href");
+
+        if (href != NULL) {
+            char *s = strdup(href);
+
+            CRM_ASSERT(s != NULL);
+            *list = g_list_append(*list, s);
+        }
+    }
+
+    freeXpathObject(xpathObj);
+    free_xml(ptr);
+}
+
+static void
+add_schema_file_to_xml(xmlNode *parent, const char *name, GList **already_included)
+{
+    char *contents = NULL;
+    xmlNode *file_node = NULL;
+    GList *includes = NULL;
+
+    /* If we already included this file, don't do so again. */
+    if (g_list_find_custom(*already_included, name, (GCompareFunc) strcmp) != NULL) {
+        return;
+    }
+
+    contents = pcmk__schema_contents(name);
+
+    if (contents == NULL) {
+        return;
+    }
+
+    /* Create a new <file path="..."> node with the contents of the file
+     * as a CDATA block underneath it.
+     */
+    file_node = create_xml_node(parent, PCMK__XA_FILE);
+
+    if (file_node == NULL) {
+        return;
+    }
+
+    if (g_str_has_suffix(name, ".rng")) {
+        crm_xml_add(file_node, PCMK__XA_PATH, name);
+        *already_included = g_list_prepend(*already_included, strdup(name));
```

Comment:
Worth doing the strdup() separately to assert non-null

Reply:
Fixed, added an assert in the else case as well, and switched to using pcmk__ends_with.